File: packages/sdk/src/harness/core/cost_controller.py
"""
Cost Controller for managing session budgets.

Prevents runaway costs by enforcing token and tool call limits.
Supports multi-level budget management: session, user, and global.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from datetime import datetime
from typing import TYPE_CHECKING, Any

from harness.types import (
    CostConfig,
    GlobalBudgetExceededError,
    ProgressEventType,
    TokenUsage,
    UserBudgetExceededError,
    UserUsage,
)

if TYPE_CHECKING:
    from harness.core.cost_storage import CostStorage
    from harness.types import ProgressCallback, ProgressEvent

logger = logging.getLogger(__name__)


@dataclass
class BudgetStatus:
    """Current budget status."""
    is_within_budget: bool
    usage: TokenUsage
    config: CostConfig
    warning_message: str | None = None
    should_compress: bool = False
    should_downgrade: bool = False
    usage_ratio: float = 0.0

    @property
    def is_warning(self) -> bool:
        """Check if in warning state."""
        return self.warning_message is not None and self.is_within_budget

    @property
    def remaining_tokens(self) -> int:
        """Tokens remaining in budget."""
        return max(0, self.config.max_tokens_per_session - self.usage.total_tokens)

    @property
    def remaining_tool_calls(self) -> int:
        """Tool calls remaining in budget."""
        return max(0, self.config.max_tool_calls_per_session - self.usage.tool_calls)


@dataclass
class UserBudgetStatus:
    """User-level budget status."""
    is_within_budget: bool
    usage: UserUsage
    config: CostConfig
    warning_message: str | None = None
    usage_ratio: float = 0.0


@dataclass
class GlobalBudgetStatus:
    """Global budget status."""
    is_within_budget: bool
    current_cost: float
    budget: float
    warning_message: str | None = None
    should_throttle: bool = False


class CostController:
    """
    Controller for managing multi-level budgets.

    Enforces limits on:
    - Session level: tokens, tool calls, iterations
    - User level: daily tokens, hourly requests
    - Global level: daily budget in USD

    Emits warnings at configurable threshold (default 80%).

    Example:
        >>> config = CostConfig(max_tokens_per_session=100_000)
        >>> controller = CostController(config)
        >>>
        >>> # Check before operation
        >>> status = controller.check(usage)
        >>> if not status.is_within_budget:
        ...     raise BudgetExceededError(status.warning_message)
    """

    def __init__(
        self,
        config: CostConfig | None = None,
        storage: "CostStorage | None" = None,
        on_progress: "ProgressCallback | None" = None,
    ):
        self.config = config or CostConfig()
        self._storage = storage
        self._on_progress = on_progress
        self._session_usage: dict[str, TokenUsage] = {}
        self._request_iterations: dict[str, int] = {}

    def check(self, usage: TokenUsage, session_id: str | None = None) -> BudgetStatus:
        """
        Check if usage is within budget.

        Args:
            usage: Current token usage
            session_id: Optional session ID for tracking

        Returns:
            BudgetStatus with check results
        """
        is_within, warning = usage.check_budget(self.config)
        usage_ratio = usage.total_tokens / self.config.max_tokens_per_session

        should_compress = False
        should_downgrade = False

        if not is_within:
            if self.config.action_on_exceed == "compress":
                should_compress = True
                is_within = True
            elif self.config.action_on_exceed == "downgrade":
                should_downgrade = True
                is_within = True

        status = BudgetStatus(
            is_within_budget=is_within,
            usage=usage,
            config=self.config,
            warning_message=warning,
            should_compress=should_compress,
            should_downgrade=should_downgrade,
            usage_ratio=usage_ratio,
        )

        # Emit progress event for warnings
        if warning and self._on_progress:
            from harness.types import ProgressEvent
            event_type = ProgressEventType.ERROR if not is_within else ProgressEventType.STATE_CHANGE
            self._on_progress(ProgressEvent(
                type=event_type,
                message=warning,
                data={
                    "usage_ratio": usage_ratio,
                    "total_tokens": usage.total_tokens,
                    "limit": self.config.max_tokens_per_session,
                },
            ))

        return status

    def check_user_budget(self, user_id: str) -> UserBudgetStatus:
        """
        Check user-level budget.

        Args:
            user_id: User identifier

        Returns:
            UserBudgetStatus with check results
        """
        if not self._storage:
            # No storage configured, skip user-level check
            return UserBudgetStatus(
                is_within_budget=True,
                usage=UserUsage(user_id=user_id),
                config=self.config,
            )

        usage = self._storage.get_user_usage(user_id)
        is_within, warning = usage.check_budget(self.config)
        usage_ratio = usage.daily_tokens / self.config.daily_token_limit if self.config.daily_token_limit > 0 else 0

        return UserBudgetStatus(
            is_within_budget=is_within,
            usage=usage,
            config=self.config,
            warning_message=warning,
            usage_ratio=usage_ratio,
        )

    def check_global_budget(self) -> GlobalBudgetStatus:
        """
        Check global budget.

        Returns:
            GlobalBudgetStatus with check results
        """
        if not self._storage:
            return GlobalBudgetStatus(
                is_within_budget=True,
                current_cost=0,
                budget=self.config.global_daily_budget_usd,
            )

        usage = self._storage.get_global_usage()
        usage_ratio = usage.daily_cost_usd / self.config.global_daily_budget_usd if self.config.global_daily_budget_usd > 0 else 0

        is_within = usage.daily_cost_usd < self.config.global_daily_budget_usd
        should_throttle = (
            self.config.auto_throttle and
            usage_ratio >= 0.8
        )

        warning = None
        if not is_within:
            warning = f"Global budget exceeded: ${usage.daily_cost_usd:.2f}/${self.config.global_daily_budget_usd:.2f}"
        elif should_throttle:
            warning = f"Global budget warning: {usage_ratio:.0%} of daily budget used"

        return GlobalBudgetStatus(
            is_within_budget=is_within,
            current_cost=usage.daily_cost_usd,
            budget=self.config.global_daily_budget_usd,
            warning_message=warning,
            should_throttle=should_throttle,
        )

    def check_all(
        self,
        usage: TokenUsage,
        session_id: str | None = None,
        user_id: str | None = None,
    ) -> tuple[BudgetStatus, UserBudgetStatus | None, GlobalBudgetStatus | None]:
        """
        Check all budget levels.

        Args:
            usage: Session token usage
            session_id: Session identifier
            user_id: User identifier (optional)

        Returns:
            Tuple of (session_status, user_status, global_status)
        """
        session_status = self.check(usage, session_id)
        user_status = self.check_user_budget(user_id) if user_id else None
        global_status = self.check_global_budget()

        return session_status, user_status, global_status

    def check_iteration(self, iteration: int, session_id: str | None = None) -> bool:
        """
        Check if iteration count is within limit.

        Args:
            iteration: Current iteration number
            session_id: Optional session ID

        Returns:
            True if within limit
        """
        if iteration >= self.config.max_iterations_per_request:
            logger.warning(
                f"Iteration limit reached: {iteration}/{self.config.max_iterations_per_request}"
            )
            return False
        return True

    def record_usage(
        self,
        session_id: str,
        input_tokens: int = 0,
        output_tokens: int = 0,
        tool_call: bool = False,
        user_id: str | None = None,
        cost_usd: float = 0.0,
    ) -> TokenUsage:
        """
        Record usage for a session.

        Args:
            session_id: Session to record for
            input_tokens: Input tokens to add
            output_tokens: Output tokens to add
            tool_call: Whether a tool call was made
            user_id: Optional user ID for user-level tracking
            cost_usd: Optional cost in USD for global tracking

        Returns:
            Updated TokenUsage for the session
        """
        if session_id not in self._session_usage:
            self._session_usage[session_id] = TokenUsage()

        usage = self._session_usage[session_id]
        usage.input_tokens += input_tokens
        usage.output_tokens += output_tokens
        if tool_call:
            usage.tool_calls += 1

        # Record user-level usage if storage is available
        if self._storage and user_id:
            self._storage.record_user_usage(
                user_id,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                request=True,
            )

        # Record global usage if storage is available
        if self._storage and cost_usd > 0:
            self._storage.record_global_usage(
                cost_usd=cost_usd,
                tokens=input_tokens + output_tokens,
            )

        return usage

    def get_session_usage(self, session_id: str) -> TokenUsage:
        """
        Get usage for a session.

        Args:
            session_id: Session ID

        Returns:
            TokenUsage for the session (or empty if not tracked)
        """
        return self._session_usage.get(session_id, TokenUsage())

    def reset_session(self, session_id: str) -> None:
        """
        Reset usage tracking for a session.

        Args:
            session_id: Session to reset
        """
        self._session_usage.pop(session_id, None)
        self._request_iterations.pop(session_id, None)

    def should_stop(self, usage: TokenUsage) -> bool:
        """
        Check if execution should stop due to budget.

        Args:
            usage: Current usage

        Returns:
            True if should stop
        """
        status = self.check(usage)
        return not status.is_within_budget and not status.should_compress

    def should_compress(self, usage: TokenUsage) -> bool:
        """
        Check if context should be compressed.

        Args:
            usage: Current usage

        Returns:
            True if compression needed
        """
        status = self.check(usage)
        return status.should_compress

    def should_downgrade(self, usage: TokenUsage) -> bool:
        """
        Check if model should be downgraded.

        Args:
            usage: Current usage

        Returns:
            True if should downgrade
        """
        status = self.check(usage)
        return status.should_downgrade

    def get_fallback_model(self) -> str:
        """Get the fallback model for budget-constrained scenarios."""
        return self.config.fallback_model

    @property
    def stats(self) -> dict[str, Any]:
        """Get controller statistics."""
        return {
            "config": {
                "max_tokens_per_session": self.config.max_tokens_per_session,
                "max_tool_calls_per_session": self.config.max_tool_calls_per_session,
                "max_iterations_per_request": self.config.max_iterations_per_request,
                "daily_token_limit": self.config.daily_token_limit,
                "hourly_request_limit": self.config.hourly_request_limit,
                "global_daily_budget_usd": self.config.global_daily_budget_usd,
                "warning_threshold": self.config.warning_threshold,
            },
            "sessions_tracked": len(self._session_usage),
            "storage_enabled": self._storage is not None,
        }

File: packages/sdk-java/harness-sdk-core/src/main/java/com/harness/core/CostController.java
package com.harness.core;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.harness.types.CostConfig;
import com.harness.types.ProgressEvent;
import com.harness.types.ProgressEventType;
import com.harness.types.TokenUsage;

/**
 * Controller for managing multi-level budgets.
 *
 * Enforces limits on:
 * - Session level: tokens, tool calls, iterations
 * - User level: daily tokens, hourly requests
 * - Global level: daily budget in USD
 *
 * Emits warnings at configurable threshold (default 80%).
 *
 * Example:
 * <pre>
 * CostConfig config = CostConfig.builder()
 *     .maxTokensPerSession(100_000)
 *     .build();
 * CostController controller = new CostController(config);
 *
 * // Check before operation
 * BudgetStatus status = controller.check(usage);
 * if (!status.isWithinBudget()) {
 *     throw new BudgetExceededException(status.getWarningMessage());
 * }
 * </pre>
 */
public class CostController {

    private static final Logger logger = LoggerFactory.getLogger(CostController.class);

    private final CostConfig config;
    private final CostStorage storage;
    private final Consumer<ProgressEvent> onProgress;

    // Session tracking
    private final Map<String, TokenUsage> sessionUsage = new HashMap<>();
    private final Map<String, Integer> requestIterations = new HashMap<>();

    // Global usage (in-memory)
    private final GlobalUsage globalUsage = new GlobalUsage();

    // User usage cache (in-memory, should be backed by storage)
    private final Map<String, UserUsage> userUsageCache = new HashMap<>();

    /**
     * Interface for cost storage backend.
     */
    public interface CostStorage {
        UserUsage getUserUsage(String userId);
        void recordUserUsage(String userId, int inputTokens, int outputTokens, boolean request);
        GlobalUsage getGlobalUsage();
        void recordGlobalUsage(double costUsd, int tokens);
    }

    public CostController() {
        this(CostConfig.defaults(), null, null);
    }

    public CostController(CostConfig config) {
        this(config, null, null);
    }

    public CostController(CostConfig config, CostStorage storage, Consumer<ProgressEvent> onProgress) {
        this.config = config;
        this.storage = storage;
        this.onProgress = onProgress;
    }

    /**
     * Check if usage is within budget.
     */
    public BudgetStatus check(TokenUsage usage, String sessionId) {
        Object[] result = usage.checkBudget(config);
        boolean isWithin = (boolean) result[0];
        String warning = (String) result[1];
        double usageRatio = (double) usage.totalTokens() / config.maxTokensPerSession();

        boolean shouldCompress = false;
        boolean shouldDowngrade = false;

        if (!isWithin) {
            if ("compress".equals(config.actionOnExceed())) {
                shouldCompress = true;
                isWithin = true;
            } else if ("downgrade".equals(config.actionOnExceed())) {
                shouldDowngrade = true;
                isWithin = true;
            }
        }

        BudgetStatus status = new BudgetStatus(
            isWithin, usage, config, warning, shouldCompress, shouldDowngrade, usageRatio
        );

        // Emit progress event for warnings
        if (warning != null && onProgress != null) {
            ProgressEventType eventType = !isWithin ? ProgressEventType.ERROR : ProgressEventType.STATE_CHANGE;
            onProgress.accept(ProgressEvent.of(
                eventType,
                warning,
                Map.of(
                    "usageRatio", usageRatio,
                    "totalTokens", usage.totalTokens(),
                    "limit", config.maxTokensPerSession()
                )
            ));
        }

        return status;
    }

    /**
     * Check user-level budget.
     */
    public UserBudgetStatus checkUserBudget(String userId) {
        UserUsage usage;
        if (storage != null) {
            usage = storage.getUserUsage(userId);
        } else {
            usage = userUsageCache.computeIfAbsent(userId, UserUsage::new);
        }

        Object[] result = usage.checkBudget(config);
        boolean isWithin = (boolean) result[0];
        String warning = (String) result[1];
        double usageRatio = config.dailyTokenLimit() > 0
            ? (double) usage.getDailyTokens() / config.dailyTokenLimit()
            : 0;

        return new UserBudgetStatus(isWithin, usage, config, warning, usageRatio);
    }

    /**
     * Check global budget.
     */
    public GlobalBudgetStatus checkGlobalBudget() {
        GlobalUsage usage = storage != null ? storage.getGlobalUsage() : globalUsage;

        double usageRatio = config.globalDailyBudgetUsd() > 0
            ? usage.getDailyCostUsd() / config.globalDailyBudgetUsd()
            : 0;

        boolean isWithin = usage.getDailyCostUsd() < config.globalDailyBudgetUsd();
        boolean shouldThrottle = config.autoThrottle() && usageRatio >= 0.8;

        String warning = null;
        if (!isWithin) {
            warning = String.format("Global budget exceeded: $%.2f/$%.2f",
                usage.getDailyCostUsd(), config.globalDailyBudgetUsd());
        } else if (shouldThrottle) {
            warning = String.format("Global budget warning: %.0f%% of daily budget used", usageRatio * 100);
        }

        return new GlobalBudgetStatus(isWithin, usage.getDailyCostUsd(), config.globalDailyBudgetUsd(),
            warning, shouldThrottle);
    }

    /**
     * Check all budget levels.
     */
    public BudgetCheckResult checkAll(TokenUsage usage, String sessionId, String userId) {
        BudgetStatus sessionStatus = check(usage, sessionId);
        UserBudgetStatus userStatus = userId != null ? checkUserBudget(userId) : null;
        GlobalBudgetStatus globalStatus = checkGlobalBudget();

        return new BudgetCheckResult(sessionStatus, userStatus, globalStatus);
    }

    /**
     * Check if iteration count is within limit.
     */
    public boolean checkIteration(int iteration, String sessionId) {
        if (iteration >= config.maxIterationsPerRequest()) {
            logger.warn("Iteration limit reached: {}/{}", iteration, config.maxIterationsPerRequest());
            return false;
        }
        return true;
    }

    /**
     * Record usage for a session.
     */
    public TokenUsage recordUsage(String sessionId, int inputTokens, int outputTokens,
                                   boolean toolCall, String userId, double costUsd) {
        TokenUsage usage = sessionUsage.computeIfAbsent(sessionId, k -> new TokenUsage());

        usage = usage.addInputTokens(inputTokens);
        usage = usage.addOutputTokens(outputTokens);
        if (toolCall) {
            usage = usage.addToolCall();
        }
        sessionUsage.put(sessionId, usage);

        // Record user-level usage
        if (storage != null && userId != null) {
            storage.recordUserUsage(userId, inputTokens, outputTokens, true);
        } else if (userId != null) {
            UserUsage userUsage = userUsageCache.computeIfAbsent(userId, UserUsage::new);
            userUsage.addTokens(inputTokens + outputTokens);
            userUsage.addRequest();
        }

        // Record global usage
        if (storage != null && costUsd > 0) {
            storage.recordGlobalUsage(costUsd, inputTokens + outputTokens);
        } else if (costUsd > 0) {
            globalUsage.addCost(costUsd);
            globalUsage.addTokens(inputTokens + outputTokens);
        }

        return usage;
    }

    /**
     * Get usage for a session.
     */
    public TokenUsage getSessionUsage(String sessionId) {
        return sessionUsage.getOrDefault(sessionId, new TokenUsage());
    }

    /**
     * Reset usage tracking for a session.
     */
    public void resetSession(String sessionId) {
        sessionUsage.remove(sessionId);
        requestIterations.remove(sessionId);
    }

    /**
     * Check if execution should stop due to budget.
     */
    public boolean shouldStop(TokenUsage usage) {
        BudgetStatus status = check(usage, null);
        return !status.isWithinBudget() && !status.shouldCompress();
    }

    /**
     * Check if context should be compressed.
     */
    public boolean shouldCompress(TokenUsage usage) {
        BudgetStatus status = check(usage, null);
        return status.shouldCompress();
    }

    /**
     * Check if model should be downgraded.
     */
    public boolean shouldDowngrade(TokenUsage usage) {
        BudgetStatus status = check(usage, null);
        return status.shouldDowngrade();
    }

    /**
     * Get the fallback model for budget-constrained scenarios.
     */
    public String getFallbackModel() {
        return config.fallbackModel();
    }

    /**
     * Get controller statistics.
     */
    public Map<String, Object> getStats() {
        Map<String, Object> configMap = new HashMap<>();
        configMap.put("maxTokensPerSession", config.maxTokensPerSession());
        configMap.put("maxToolCallsPerSession", config.maxToolCallsPerSession());
        configMap.put("maxIterationsPerRequest", config.maxIterationsPerRequest());
        configMap.put("dailyTokenLimit", config.dailyTokenLimit());
        configMap.put("hourlyRequestLimit", config.hourlyRequestLimit());
        configMap.put("globalDailyBudgetUsd", config.globalDailyBudgetUsd());
        configMap.put("warningThreshold", config.warningThreshold());

        Map<String, Object> stats = new HashMap<>();
        stats.put("config", configMap);
        stats.put("sessionsTracked", sessionUsage.size());
        stats.put("storageEnabled", storage != null);
        return stats;
    }

    /**
     * Result of checking all budget levels.
     */
    public record BudgetCheckResult(
        BudgetStatus sessionStatus,
        UserBudgetStatus userStatus,
        GlobalBudgetStatus globalStatus
    ) {}
}
